Import: add DXF statistics reporter

furgo16 opened the pull request:
Add a console DXF statistics reporter for the FreeCAD C++ importer. Help both users and developers understand the content of DXF files and the importer's behaviour.

> [!NOTE]
> This PR does not change the functionality of the DXF importer. It simply adds reporting capabilities to it.

Goals:
- Provide more feedback to the user. DXF import is a complex process: there are two importers currently and each supports a set of features. More information removes uncertainty from the process.
- Provide more context for reporting bugs. In particular, the import settings have been included in the report, so that they can be pasted on issue reports.
- Provide the user more context on what the optimal import settings are. There are currently multiple settings and setting combinations that lead to different importer results. Comparing the report output with different settings can guide the user to find the most optimal combination for their goals. 

Reports on:
- DXF file metadata (file name, DXF version, file encoding, scale)
- Effective scaling applied to the document depending on import settings
- Time to import
- Import settings
- Individual read entity counts, with an indicator of whether they are supported or not.
- Total entities read and total FreeCAD objects created
- Unsupported features/entities count (*), with details on line no. and DXF handle for each. These are generally skipped by the importer.

(*) Note: the original code did report on this, but as a warning that always popped up on import. To have a less disruptive import operation, these original warnings have been removed. Instead, the counts are included in the report.

<details><summary>Expand for a typical listing from the reporter:</summary>

```
--- DXF import summary ---
Import of file: '/home/me/sample-arch-dxf.dxf'

DXF version: AC1027
File encoding: utf_8
File units: Millimeters (from $INSUNITS)
Manual scaling factor: 1.000000
Final scaling: 1 DXF unit = 1.0000 mm

Performance:
  - C++ import time: 2.3620 seconds
  - Total import time: 2.8031 seconds

Import settings:
  - Import hidden blocks: No
  - Import layout objects: No
  - Import points: No
  - Import texts and dimensions: No
  - Join geometry: No
  - Manual scaling factor: 1.000000
  - Merge option: Create Part shapes
  - Use colors from the DXF file: Yes
  - Use layers: Yes

Entity counts:
  - ARC: 578
  - ATTDEF: 4 (*)
  - BLOCK: 159
  - CIRCLE: 37
  - DIMENSION: 31
  - ELLIPSE: 21
  - ENTITIES_IN_PAPERSPACE: 105
  - HATCH: 8 (*)
  - INSERT: 24
  - LINE: 2921
  - LWPOLYLINE: 187
  - MLINE: 5 (*)
  - REGION: 1 (*)
  - SPLINE: 90
  - TEXT: 9
  - VIEWPORT: 2 (*)
----------------------------
  Total entities read: 4182
FreeCAD objects created: 684

(*) Entity type not supported by importer.

Unsupported features:
  - Entity type 'ATTDEF': 4 time(s) at line 9488 (handle D3), line 9530 (handle D4), line 9572 (handle D5), line 9614 (handle D6)
  - Entity type 'HATCH': 8 time(s). Examples: line 42498 (handle 2843), line 50930 (handle 349A), line 148756 (handle 111B1), line 148922 (handle 111B2), line 149082 (handle 111B3), ...
  - Entity type 'MLINE': 5 time(s) at line 147470 (handle 10D3D), line 147576 (handle 10D3E), line 148438 (handle 10DD0), line 148544 (handle 10DD1), line 148650 (handle 10DD2)
  - Entity type 'REGION': 1 time(s) at line 14880 (handle 654)
  - Entity type 'VIEWPORT': 2 time(s) at line 155372 (handle FB23), line 155492 (handle FB25)
--- End of summary ---

```

</details>

Potential future improvements and features:

1. Report on a dialog if the GUI is up (will probably be added in a subsequent PR)
2.  Report on skipped entities and the reason for skipping them
3. Better reporting on anonymous vs user blocks (added in a separate PR: https://github.com/FreeCAD/FreeCAD/pull/22045)
4. Review output formatting

Known limitations:
- There is no distinction between user and system (anonymous) BLOCK entities when reporting. This is fixed in https://github.com/FreeCAD/FreeCAD/pull/22045
- FreeCAD object count is not correct for Compound objects (only the compound is called and not their children). This is fixed in https://github.com/FreeCAD/FreeCAD/pull/22045

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->

furgo16 commented:
The failing tests in the MacOS pixi build seem unrelated to this PR:

```
ImportError: /Users/runner/work/FreeCAD/FreeCAD/.pixi/envs/default/Mod/Draft/drafttests/test_modification.py(606)<class 'ImportError'>: dlopen(/Users/runner/work/FreeCAD/FreeCAD/.pixi/envs/default/lib/TechDraw.so, 0x0002): symbol not found in flat namespace '__ZN8CDxfRead18UnsupportedFeatureIJEEEvPKcDpOT_'
```

furgo16 commented:
I might need some help, as I can't really figure out what the issue with the build failures on the MacOS build is. It's not that the tests are not passing, some tests are failing to run.

furgo16 commented:
> I would like preserve the display of the line number of the first occurrence of each unsupported-feature message.
This has been a very useful feature both in terms of developing the importer and also in pinpointing issues with the imported DXF file contents.

Thanks for the review. I had actually thought about this as well, but in other terms: adding the handle for the supported entity in the report. I guess either approach would work.

The only reason I did not add (or re-add) this was scope and effort. This PR already grew too big and I wanted to get it done rather than perfect. In terms of time it's personally a significant undertaking for me to switch from Python to C++, where I'm by no means proficient. So from that point of view, I also wanted to get the PR done :)

Would it be an acceptable compromise to merge as-is and add the unsupported entities line or handle in a follow-up PR?

kpemartin commented:
I guess that would be OK, as long as there is some sort of reminder to do it. A TODO comment in the code? a new Issue? I'm not sure what is appropriate for these little reminders.

Also, unsupported features are not necessarily entities in the ENTITY or BLOCK constructs, so there is not necessarily a handle to display. Trying to find part of the DXF by handle in a text editor requires a 2-line search pattern, to look for the Handle tag (I don't recall what it is) followed by the handle itself on the next line.

Just going to a numbered line works better for both cases.

Though I can see where the Handle, if available, might be useful to track down the origin of the unsupported feature in whatever program generated the DXF in the first place. Maybe both should be there, if available, or some other identifying information. Ah, isn't feature creep wonderful?

furgo16 commented:
I've now added the (line, handle) pair reporting. If there's no handle associated, the report will indicate so. As before, only a maximum of 5 unsupported entries with line, handle pair per entity will be reported. The idea is to report a subset only, in order not to overflow the console output. The final count will still include all unsupported instances.

While at it, I've added a couple of small improvements to the report:
- Filename of the DXF being processed
- On the processed entities count, an indicator shows which ones are NOT supported 

Here's a sample report output from reading this [DXF sample file](https://github.com/user-attachments/files/20880953/sample-arch-dxf.zip) (kindly provided by @semhustej):

<details>

```
--- DXF import summary ---
Import of file: '/home/me/sample-arch-dxf.dxf'

DXF version: AC1027
File encoding: utf_8
File units: Millimeters (from $INSUNITS)
Manual scaling factor: 1.000000
Final scaling: 1 DXF unit = 1.0000 mm

Performance:
  - C++ import time: 2.3620 seconds
  - Total import time: 2.8031 seconds

Import settings:
  - Import hidden blocks: No
  - Import layout objects: No
  - Import points: No
  - Import texts and dimensions: No
  - Join geometry: No
  - Manual scaling factor: 1.000000
  - Merge option: Create Part shapes
  - Use colors from the DXF file: Yes
  - Use layers: Yes

Entity counts:
  - ARC: 578
  - ATTDEF: 4 (*)
  - BLOCK: 159
  - CIRCLE: 37
  - DIMENSION: 31
  - ELLIPSE: 21
  - ENTITIES_IN_PAPERSPACE: 105
  - HATCH: 8 (*)
  - INSERT: 24
  - LINE: 2921
  - LWPOLYLINE: 187
  - MLINE: 5 (*)
  - REGION: 1 (*)
  - SPLINE: 90
  - TEXT: 9
  - VIEWPORT: 2 (*)
----------------------------
  Total entities read: 4182
FreeCAD objects created: 684

(*) Entity type not supported by importer.

Unsupported features:
  - Entity type 'ATTDEF': 4 time(s) at line 9488 (handle D3), line 9530 (handle D4), line 9572 (handle D5), line 9614 (handle D6)
  - Entity type 'HATCH': 8 time(s). Examples: line 42498 (handle 2843), line 50930 (handle 349A), line 148756 (handle 111B1), line 148922 (handle 111B2), line 149082 (handle 111B3), ...
  - Entity type 'MLINE': 5 time(s) at line 147470 (handle 10D3D), line 147576 (handle 10D3E), line 148438 (handle 10DD0), line 148544 (handle 10DD1), line 148650 (handle 10DD2)
  - Entity type 'REGION': 1 time(s) at line 14880 (handle 654)
  - Entity type 'VIEWPORT': 2 time(s) at line 155372 (handle FB23), line 155492 (handle FB25)
--- End of summary ---
```

</details>

furgo16 commented:
Addressing the rest of comments I didn't get to reply the last time (sorry for the delay),

> I'm wondering if we shouldn't try more to keep dxf.cpp clean and do more of the stats processing out of it, but at the same time there is little point in maintaining dxf.cpp pristine any more, there is a long time that it's only ours now...

I tried to keep the parser pristine, but there is no way some of the changes in this PR (and subsequent ones I'm working on) can be done solely on the upper layers. Or at least doing so is beyond my capabilities. The changes, while not minimal, are not huge either, and are self-contained.

I did some research, and it seems the DXF parser code comes from https://github.com/danheeks/dxfconv, specifically the `dxf.cpp`/`dxf.h` pair only. The cleanest approach might be to set up a git submodule, but:

- I'm not sure if the submodule approach is viable when sharing only a subset (two) of the original repo's files.
- I'm also uncertain whether a) the upstream repo is still maintained and b) whether our changes would be interesting for upstream at all.
- The code is in git, so all history and changes are traceable. Tracking the common ancestor between our code and upstream should be doable in any case.
- For the most part the parser works, but if we were to put more work into it, rather than extending, I'd look into an architecture that's more inline with some of the other importers in FreeCAD, which (I _think_, I have just glossed over them) use [OCCT's XCAF/XDE](https://dev.opencascade.org/doc/overview/html/occt_user_guides__xde.html) model. I'm not an expert, so I might be wrong here.

> Going forward this will require deciding, for new messages, whether the message is sufficiently important to be reported directly to the console, or if it can just be added to the stats (either in an existing property/collection or as a new property). The risk in the latter case is that the import may occur in a context where the stats are not reported, and this could result in the loss of an important message that could explain some unexpected import behaviour.

Good point. At this time I only moved the unsupported features messages to the stats reporter as a central place for this information, as I thought they would belong there as a notice to the user that their model might not be imported in high fidelity. The rest of warnings are still active and independent from the reporter.
